wsagent/che-core-api-languageserver/src/main/java/org/eclipse/che/api/languageserver/service/TextDocumentService.java

```diff
@@ -776,6 +800,115 @@ public boolean handleResult(
+          new LSOperation<InitializedLanguageServer, WorkspaceEdit>() {
+            @Override
+            public boolean canDo(InitializedLanguageServer server) {
+              Boolean renameProvider =
+                  server.getInitializeResult().getCapabilities().getRenameProvider();
+              return renameProvider != null && renameProvider;
+            }
+
+            @Override
+            public CompletableFuture<WorkspaceEdit> start(InitializedLanguageServer element) {
+              return element.getServer().getTextDocumentService().rename(renameParams);
+            }
+
+            @Override
+            public boolean handleResult(InitializedLanguageServer element, WorkspaceEdit result) {
+
+              addRenameResult(edits, element.getLauncher().getDescription().getId(), result);
+              return true;
+            }
+          };
+      OperationUtil.doInParallel(servers, op, TimeUnit.SECONDS.toMillis(30));
+    } catch (LanguageServerException e) {
+      throw new JsonRpcException(-27000, e.getMessage());
+    }
+    return new RenameResultDto(new RenameResult(edits));
+  }
+
+  private void addRenameResult(
+      Map<String, ExtendedWorkspaceEdit> map, String id, WorkspaceEdit workspaceEdit) {
+
+    ExtendedWorkspaceEdit result = new ExtendedWorkspaceEdit();
+    List<ExtendedTextDocumentEdit> edits = new ArrayList<>();
+    if (workspaceEdit.getDocumentChanges() != null) {
+      for (TextDocumentEdit documentEdit : workspaceEdit.getDocumentChanges()) {
+        ExtendedTextDocumentEdit edit = new ExtendedTextDocumentEdit();
+        edit.setTextDocument(documentEdit.getTextDocument());
+        edit.getTextDocument().setUri(removePrefixUri(edit.getTextDocument().getUri()));
+        edit.setEdits(
+            convertToExtendedEdit(
+                documentEdit.getEdits(), removeUriScheme(documentEdit.getTextDocument().getUri())));
+        edits.add(edit);
+      }
+    } else if (workspaceEdit.getChanges() != null) {
+      for (Entry<String, List<TextEdit>> entry : workspaceEdit.getChanges().entrySet()) {
+        ExtendedTextDocumentEdit edit = new ExtendedTextDocumentEdit();
+        VersionedTextDocumentIdentifier documentIdentifier = new VersionedTextDocumentIdentifier();
+        documentIdentifier.setVersion(-1);
+        documentIdentifier.setUri(removePrefixUri(entry.getKey()));
+        edit.setTextDocument(documentIdentifier);
+        edit.setEdits(convertToExtendedEdit(entry.getValue(), removeUriScheme(entry.getKey())));
+        edits.add(edit);
+      }
+    }
+
+    if (!edits.isEmpty()) {
+      result.setDocumentChanges(edits);
+      map.put(id, result);
+    }
+  }
+
+  private List<ExtendedTextEdit> convertToExtendedEdit(List<TextEdit> edits, String filePath) {
+    try {
+      // for some reason C# LS sends ws related path,
+      if (!isStartWithProject(filePath)) {
+        filePath = prefixProject(filePath);
+      }
+      String fileContent =
```

Comment:
Sorry to chime in late, but since we'll face a similar problem with the "references" for the jdt.ls, etc. I think this approach will not scale for large sets of changes. We should be thinking towards a UI where we fetch the concrete line changes ("snippets") when we need them.